[SPARK-30246][CORE] OneForOneStreamManager might leak memory in connectionTerminated

hensg opened the pull request:
### What changes were proposed in this pull request?

Ensure that all StreamStates are removed from OneForOneStreamManager memory map even if there's an error trying to release buffers

### Why are the changes needed?

OneForOneStreamManager may not remove all StreamStates from memory map when a connection is terminated. A RuntimeException might be thrown in StreamState$buffers.next() by one of ExternalShuffleBlockResolver$getBlockData... **breaking the loop through streams.entrySet(), keeping StreamStates in memory forever leaking memory.**
That may happen when an application is terminated abruptly and executors removed before the connection is terminated or if shuffleIndexCache fails to get ShuffleIndexInformation

References:
https://github.com/apache/spark/blob/ee050ddbc6eb6bc08c7751a0eb00e7a05b011b52/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java#L319

https://github.com/apache/spark/blob/ee050ddbc6eb6bc08c7751a0eb00e7a05b011b52/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java#L357

https://github.com/apache/spark/blob/ee050ddbc6eb6bc08c7751a0eb00e7a05b011b52/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java#L195

https://github.com/apache/spark/blob/ee050ddbc6eb6bc08c7751a0eb00e7a05b011b52/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java#L208

https://github.com/apache/spark/blob/ee050ddbc6eb6bc08c7751a0eb00e7a05b011b52/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java#L330


### Does this PR introduce any user-facing change?
No


### How was this patch tested?
Unit test added

hensg committed:
OneForOneStreamManager might leak memory in connectionTerminated

OneForOneStreamManager may not remove all StreamStates from memory map when a connection is terminated.

A RuntimeException might be thrown in StreamState$buffers.next() by one of ExternalShuffleBlockResolver$getBlockData/getRddBlock... breaking the loop through streams.entrySet()

This commit removes all StreamStates from memory map first and, after that, releases all buffers from each removed StreamState

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

HeartSaVioR commented on common/network-common/src/main/java/org/apache/spark/network/server/OneForOneStreamManager.java:
```diff
@@ -117,18 +118,21 @@ public static String genStreamChunkId(long streamId, int chunkId) {
+    }
+
+    for (StreamState state: removedStates) {
+      // Release all remaining buffers.
```
Since we're here, would we be better to try our best to clean up everything? Like adding try-catch and try to release buffers for all removedStates even there's any exception on releasing any state. 

Regarding throwing exception, we could just store the first exception if anything occurs (other exceptions could be just logged) and throw it. Actually given the implementation of `TransportRequestHandler.channelInactive`, we can even log these exceptions as ERROR and don't throw any; effectively same.

hensg replied:
What about log these exceptions as WARN?

HeartSaVioR replied:
Given TransportRequestHandler.channelInactive logs the exception as ERROR I would suggest do the same, but committers can weigh it in any way while reviewing so wouldn't be a big deal for now.

hensg replied:
If iterator `.next()` throws an exception and we `catch` it, the iterator cursor/index won't be increased and we'll be stuck at that iterator position, and the next time that `.next()` get called it will throw the same exception.

HeartSaVioR replied:
What I actually suggested was like:

```
for (StreamState state: removedStates) {
  try {
      while (state.buffers.hasNext()) {
        ManagedBuffer buffer = state.buffers.next();
        if (buffer != null) {
          buffer.release();
        }
      }
  } catch {
    case NonFatal(e) => logError("...proper log message with e...")
  }
}
```

instead of looping with hasNext() & next() even there's exception.

But looks like it also makes sense if and only if state.buffers.hasNext() would throw exception if state.buffers.next() throws exception. Then we can determine when to give up iterating.

hensg replied:
got it :man_facepalming:

012huang replied:
there exists an case, if the application has already finished(failed), executors' info of the app has been cleaned, but the connectionTerminated be called delay, when execute this code, it will look up the buffer with `appId` and `execId` , so the buffer cannot be got and released and still stay remain. this happend in spark 2.4.3 as I report and I also make a pr, pls help review(https://github.com/apache/spark/pull/27064), thanks

hensg committed:
Add try-catch block to release buffers for all removedStates even there's any exception on releasing any state.

HeartSaVioR commented:
cc. @viirya

viirya commented on common/network-common/src/main/java/org/apache/spark/network/server/OneForOneStreamManager.java:
```diff
@@ -117,19 +118,26 @@ public static String genStreamChunkId(long streamId, int chunkId) {
+      // Release all remaining buffers.
+      try {
         while (state.buffers.hasNext()) {
           ManagedBuffer buffer = state.buffers.next();
```
Once we got RuntimeException, don't we just fail? Is there memory leak?

HeartSaVioR replied:
https://github.com/apache/spark/blob/1b0570c6af25849d46ddd2d108d2fa949616b756/common/network-common/src/main/java/org/apache/spark/network/server/TransportRequestHandler.java#L87-L97

`TransportRequestHandler.channelInactive` will log error message and swallow the exception, so no, we don't fail because of exception thrown here.

viirya replied:
oh I see. Then this change may also swallow the RuntimeException so `channelInactive` cannot log as before? Should we rethrow the same exception?

HeartSaVioR replied:
So I think there're two options here with achieving goal;

1. store any exception between exceptions (as we'll try to release from all the streams) and rethrow, then channelInactive will log it. other exceptions should be logged as well but maybe at here (or a new exception containing all exceptions and throw it instead).

2. catch them, and log and swallow at here.

Technically the change from logging side would be logger name which may not a big deal, but it's also valid concern if we think channelInactive is the one to decide how to handle the exception.

viirya replied:
yea, I think it might make the caller (channelInactive or others if any) thinks everything is fine inside `connectionTerminated`. Maybe not big deal for now, but sounds a potential concern.

We might have multiple exception during releasing buffers. We can rethrow a `RuntimeException` if any exception happens during releasing buffers.

HeartSaVioR replied:
OK thanks for the input. @hensg Could you follow up the comment above? Thanks!

viirya commented:
ok to test.

SparkQA commented:
**[Test build #116058 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116058/testReport)** for PR 27064 at commit [`a9ebe4f`](https://github.com/apache/spark/commit/a9ebe4ffe15bf1e438e8f27a2b7344c4f7b6eaf7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20850/
Test PASSed.

SparkQA commented:
**[Test build #116058 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116058/testReport)** for PR 27064 at commit [`a9ebe4f`](https://github.com/apache/spark/commit/a9ebe4ffe15bf1e438e8f27a2b7344c4f7b6eaf7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116058/
Test PASSed.

012huang reviewed (commented):
there exists an case, if the application has already finished(failed), executors' info of the app has been cleaned, but the connectionTerminated be called delay, when execute this code, it will look up the buffer with `appId` and `execId` , so the buffer cannot be got and released. this happend in spark 2.4.3 as I report and I also make a pr, pls help review(https://github.com/apache/spark/pull/27060), thanks

HeartSaVioR commented:
@012huang I feel it's orthogonal to this. Could you please cc. to some committers modified around the code having bug in your PR?

HeartSaVioR reviewed (approved):
LGTM. cc. @viirya Could you please take a look again? Thanks in advance.

hensg committed:
Rethrow exception when fails to release all StreamState buffers

SparkQA commented:
**[Test build #116383 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116383/testReport)** for PR 27064 at commit [`410651d`](https://github.com/apache/spark/commit/410651d87c123f7a22f2950384d972bb7e71f30a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
**[Test build #116383 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116383/testReport)** for PR 27064 at commit [`410651d`](https://github.com/apache/spark/commit/410651d87c123f7a22f2950384d972bb7e71f30a).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116383/
Test FAILed.

hensg committed:
Fix linter linelength error

[ERROR] src/test/java/org/apache/spark/network/server/OneForOneStreamManagerSuite.java:[123] (sizes) LineLength: Line is longer than 100 characters (found 106).

SparkQA commented:
**[Test build #116384 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116384/testReport)** for PR 27064 at commit [`ac58798`](https://github.com/apache/spark/commit/ac587989396f88bb2a84064f5906b94d086bf12d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21175/
Test PASSed.

SparkQA commented:
**[Test build #116384 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116384/testReport)** for PR 27064 at commit [`ac58798`](https://github.com/apache/spark/commit/ac587989396f88bb2a84064f5906b94d086bf12d).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116384/
Test FAILed.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #116398 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116398/testReport)** for PR 27064 at commit [`ac58798`](https://github.com/apache/spark/commit/ac587989396f88bb2a84064f5906b94d086bf12d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21189/
Test PASSed.

SparkQA commented:
**[Test build #116398 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116398/testReport)** for PR 27064 at commit [`ac58798`](https://github.com/apache/spark/commit/ac587989396f88bb2a84064f5906b94d086bf12d).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116398/
Test FAILed.

viirya commented:
retest this please

SparkQA commented:
**[Test build #116414 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116414/testReport)** for PR 27064 at commit [`ac58798`](https://github.com/apache/spark/commit/ac587989396f88bb2a84064f5906b94d086bf12d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21203/
Test PASSed.

SparkQA commented:
**[Test build #116414 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116414/testReport)** for PR 27064 at commit [`ac58798`](https://github.com/apache/spark/commit/ac587989396f88bb2a84064f5906b94d086bf12d).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116414/
Test FAILed.

viirya commented:
looks like a valid failure `test_memory_limit (pyspark.tests.test_worker.WorkerMemoryTest)`? It continues to fail on this.

HeartSaVioR commented:
@viirya The test fails from other PRs as well - please refer  https://github.com/apache/spark/pull/26201#issuecomment-572598097

Looks like other PRs succeed to build because the pyspark test seems run conditionally.

viirya commented:
@HeartSaVioR hmm, I just ran the test locally and it passed.

HeartSaVioR commented:
Yeah same for me as well, but I don't have pypy. Installing now.

HeartSaVioR commented:
Oh wait I'm seeing failure "after" installing "resource", but the test fails elsewhere which is odd.

```
Running PySpark tests. Output is in /Users/jlim/WorkArea/ScalaProjects/spark/python/unit-tests.log
Will test against the following Python executables: ['/usr/local/bin/python3', 'python2.7', 'pypy']
Will test the following Python tests: ['pyspark.tests.test_worker']
Starting test(/usr/local/bin/python3): pyspark.tests.test_worker
Starting test(pypy): pyspark.tests.test_worker
Starting test(python2.7): pyspark.tests.test_worker
test_memory_limit (pyspark.tests.test_worker.WorkerMemoryTest) ... Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
/Users/jlim/WorkArea/ScalaProjects/spark/python/pyspark/context.py:219: DeprecationWarning: Support for Python 2 and Python 3 prior to version 3.6 is deprecated as of Spark 3.0. See also the plan for dropping Python 2 support at https://spark.apache.org/news/plan-for-dropping-python-2-support.html.
  DeprecationWarning)
[Stage 0:>                                                          (0 + 1) / 1]Current mem limits: 9223372036854775807 of max 9223372036854775807

Setting mem limits to 1048576 of max 1048576

/Users/jlim/WorkArea/ScalaProjects/spark/python/pyspark/tests/test_worker.py:190: RuntimeWarning: Parent module 'pyspark.tests' not found while handling absolute import
  import resource
ok
test_reuse_worker_of_parallelize_xrange (pyspark.tests.test_worker.WorkerReuseTest) ... FAIL
test_accumulator_when_reuse_worker (pyspark.tests.test_worker.WorkerTests) ... ok
test_after_exception (pyspark.tests.test_worker.WorkerTests) ... ok
test_after_jvm_exception (pyspark.tests.test_worker.WorkerTests) ... ok
test_cancel_task (pyspark.tests.test_worker.WorkerTests) ... /Users/jlim/WorkArea/ScalaProjects/spark/python/pyspark/tests/test_worker.py:45: RuntimeWarning: Parent module 'pyspark.tests' not found while handling absolute import
  import os
/Users/jlim/WorkArea/ScalaProjects/spark/python/pyspark/tests/test_worker.py:46: RuntimeWarning: Parent module 'pyspark.tests' not found while handling absolute import
  import time
ok
test_python_exception_non_hanging (pyspark.tests.test_worker.WorkerTests) ... ok
test_reuse_worker_after_take (pyspark.tests.test_worker.WorkerTests) ... ok
test_with_different_versions_of_python (pyspark.tests.test_worker.WorkerTests) ... ok

======================================================================
Finished test(/usr/local/bin/python3): pyspark.tests.test_worker (15s)
FAIL: test_reuse_worker_of_parallelize_xrange (pyspark.tests.test_worker.WorkerReuseTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jlim/WorkArea/ScalaProjects/spark/python/pyspark/tests/test_worker.py", line 176, in test_reuse_worker_of_parallelize_xrange
    self.assertTrue(pid in previous_pids)
AssertionError: False is not true

----------------------------------------------------------------------
Ran 9 tests in 15.360s

FAILED (failures=1)
```

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21240/
Test PASSed.

SparkQA commented:
**[Test build #116453 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116453/testReport)** for PR 27064 at commit [`ac58798`](https://github.com/apache/spark/commit/ac587989396f88bb2a84064f5906b94d086bf12d).

SparkQA commented:
**[Test build #116453 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116453/testReport)** for PR 27064 at commit [`ac58798`](https://github.com/apache/spark/commit/ac587989396f88bb2a84064f5906b94d086bf12d).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116453/
Test FAILed.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #116472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116472/testReport)** for PR 27064 at commit [`ac58798`](https://github.com/apache/spark/commit/ac587989396f88bb2a84064f5906b94d086bf12d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21259/
Test PASSed.

SparkQA commented:
**[Test build #116472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116472/testReport)** for PR 27064 at commit [`ac58798`](https://github.com/apache/spark/commit/ac587989396f88bb2a84064f5906b94d086bf12d).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116472/
Test FAILed.

hensg committed:
Remove extra blank line

SparkQA commented:
**[Test build #116556 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116556/testReport)** for PR 27064 at commit [`d76483f`](https://github.com/apache/spark/commit/d76483f4e3a3803a865c614d89a3233ed18f72e7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21342/
Test PASSed.

SparkQA commented:
**[Test build #116556 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116556/testReport)** for PR 27064 at commit [`d76483f`](https://github.com/apache/spark/commit/d76483f4e3a3803a865c614d89a3233ed18f72e7).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116556/
Test FAILed.

hensg commented:
retest this please

SparkQA commented:
**[Test build #116663 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116663/testReport)** for PR 27064 at commit [`b02e48a`](https://github.com/apache/spark/commit/b02e48a1b1d1f1b5409fbd86ed8ffd3e968ddf5d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21442/
Test PASSed.

SparkQA commented:
**[Test build #116663 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116663/testReport)** for PR 27064 at commit [`b02e48a`](https://github.com/apache/spark/commit/b02e48a1b1d1f1b5409fbd86ed8ffd3e968ddf5d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116663/
Test PASSed.

viirya commented on common/network-common/src/main/java/org/apache/spark/network/server/OneForOneStreamManager.java:
```diff
@@ -117,21 +117,32 @@ public static String genStreamChunkId(long streamId, int chunkId) {
+              buffer.release();
+            }
           }
+        } catch (RuntimeException e) {
```
Is this the only exception type we need to deal with here?

HeartSaVioR replied:
Other kind of exception would be Error which `TransportRequestHandler.channelInactive` is also not catching it and let the process fail.

viirya reviewed (approved):
looks good. cc @cloud-fan

cloud-fan commented on common/network-common/src/main/java/org/apache/spark/network/server/OneForOneStreamManager.java:
```diff
@@ -117,21 +117,32 @@ public static String genStreamChunkId(long streamId, int chunkId) {
       }
     }
+
+    if (failedToReleaseBuffers) {
```
instead of using a boolean flag, can we keep the actual exception (we can just keep the first one) and rethrow it? then we can retain the stack trace.

hensg committed:
Store first exception and re-throw it

SparkQA commented:
**[Test build #116727 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116727/testReport)** for PR 27064 at commit [`0fde4c9`](https://github.com/apache/spark/commit/0fde4c94fd67f6be55ee6eff511e9d046fd3cd5c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21502/
Test PASSed.

SparkQA commented:
**[Test build #116727 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116727/testReport)** for PR 27064 at commit [`0fde4c9`](https://github.com/apache/spark/commit/0fde4c94fd67f6be55ee6eff511e9d046fd3cd5c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116727/
Test PASSed.

dongjoon-hyun commented:
cc @vanzin , too.

vanzin commented:
Merging to master / 2.4.

vanzin commented:
FYI: fixed a trivial conflict in 2.4, and had to add a couple of things to the test:

- import the Assert class
- add a couple of `@SuppressWarnings` otherwise compilation was failing. Not sure why master did not complain...

dongjoon-hyun commented:
Thank you all!
